add Placement New

WalterBright opened the pull request:
This implements Placement New, described in https://github.com/WalterBright/documents/blob/master/placementnew.md

WalterBright committed:
add Placement New

WalterBright commented:
I got it to work for some basic cases. Next comes extending it to more complex ones.

WalterBright commented:
Placement new for structs seem to be working now!

UplinkCoder commented on compiler/src/dmd/backend/x86/cod2.d:
```diff
@@ -4664,7 +4664,7 @@ void cdrelconst(ref CGstate cg, ref CodeBuilder cdb,elem *e,ref regm_t pretregs)
         case TYcldouble:
             tym = TYnptr;               // don't confuse allocreg()
-            if (pretregs & (mES | mCX) || e.Ety & mTYfar)
+            if (I16 && pretregs & (mES | mCX) || e.Ety & mTYfar)
```
Looks like this is unrelated?

WalterBright replied:
yes, I was going to make it a separate PR. It's a problem that's been there forever, but it doesn't seem to negatively affect code execution

WalterBright replied:
https://github.com/dlang/dmd/pull/17065

Connor-GH commented:
> If one desires to use classes without the GC, such as in BetterC, it's just awkward to use emplace

If i'm reading this right, this will allow classes in betterC, using a kind of "allocator argument" like in $OTHER_LANGUAGES?

WalterBright commented:
Placement works for class objects now!

TurkeyMan commented:
Did you tweak it to receive pointers as arguments? I've sent several emails recently and they've all gone silent; I'm wondering if they're getting lost. I also tried commenting on the code in GitHub.

WalterBright commented:
I haven't received emails about this. I will check my spam filter. I don't know why you cannot comment on github? I thought it was you who had gone silent! Glad that's not the case.

I stewed about the placement expression as pointer for a while, and figured that making it essentially a `ref` parameter is more the D way of eschewing pointers in favor of ref. A ref also includes bounds on the size, making that check something the language can do.

Of course, placement new will always be unsafe, but at least we can prevent memory overruns with ref parameters.

TurkeyMan commented:
Weird, I've sent several messages in various ways.

Nar, it's really not appropriate as a ref.
I think the best way to think about it from a spec perspective is in terms of lifetime; placement new returns a new T, prior to the placement new call, it is NOT a T, it is just raw memory. 
I would even make the case that placement new should receive void* and not T*; that would be uniform work the class case, but it would also make the semantics very clear, prior to the call, you are handling a pointer to memory, after the new call, you receive a T as result; the lifetime of T _begins_ with the call to new, and nothing in the expression gives the impression that T pre-exists the call to new. 

I feel it's absolutely inappropriate to hand it a ref to a T to be initialized... that certainly implies that the T pre-exists. And also the asymmetry with the class case really bugs me.

Your error message "new must receive an lvalue" gives the wrong impression; an lvalue in my mind suggests that the value exists, but before the new, there is no value, just vacant memory.

I think I've convinced myself while writing this that even T* which I initially felt attracted to is wrong, and placement new should only receive void* as input... that correctly expresses the operation.

WalterBright commented:
Glad your review is getting through this time. Some good thoughts!

1. `void*` certainly works from a pragmatic, to-the-metal approach. It does come with a price, though. The compiler is helpless in diagnosing simple mistakes, like the amount of storage pointed to being insufficient. When developing a test case, I promptly made that mistake in not having sufficient size for a class instance. Fortunately, the compiler hit back with "Walter, you ignorant scum. The size is too small!" Properly chastened, I was grateful to get the message rather than baffling memory overflow bugs. It reminds me of when I resisted Andrei's suggestion that array bounds checking be always turned on, unless in `@system` code and a compiler flag was thrown. History has proven Andrei to be right. P.S. I hear C++ std::vector is going to get array bounds checking on by default! Only 25 years late!!
2. My main use of emplace in the compiler is there's a union of various Expressions in the ctfe interpreter. The placement new fits in perfectly with that, as the union object will always have sufficient size. The `new(u) AddExp` just goes perfectly with that.
3. The lifetime of any pre-existing lvalue is going to be unceremoniously terminated when used by placement new, without passing Go or collecting $200 or running the destructor. This is why it is fundamentally unsafe. But at least refs can be statically checked for correct size!
4. Maybe we can improve the lifetime support in the future. But we won't be able to do much of anything with `void*` as an argument.

WalterBright commented:
P.S. lifetime analysis of a pointer to an object is more difficult than lifetime analysis of an object, because of the possibility of multiple pointers to the same object.

TurkeyMan commented:
There is no lifetime before the call to new. The lifetime begins with new, and before new, there is only a pointer to bytes in memory. The placement new expression might in the future have integrated into it some initiation of lifetime for any future lifetime analysis; it's one of the reasons I think it's important to introduce it and concentrate initialisation in an intrinsic expression.

Your particular case with a union is obviously legit, but I assure you it's an edge case; every other case where you need placement new is when calling malloc() (or some similar allocator) to get memory, and then emplacing runtime values into that allocation.

If you're not comfortable with void*, then consider void[] which you can use for bound checking.

Your comment about lifetime analysis on any object leading IN to the new() expression is not something that makes sense to me... there is no lifetime of anything to speak of prior to the new... that's exactly what new means; it takes raw memory and turns it into live values.

The union case is actually a sort of aberration if anything. Personally I don't much like unions fundamentally from a conceptual point of view; because they express an aggregate of values despite all, or all-but-one of those values are completely invalid.
You might recall me arguing 5-10 years ago when we first ever started talking about @safe that I felt unions are absolutely fundamentally not @safe, and it's completely improper for @safe code to make contact with a union... I was shot down by several people (who are all wrong), and I still believe that's true today, and this kinda demonstrates a facet of my discomfort; the idea you are supplying an lvalue (ref) of what appears to be a _value_ to a `new()`, except it's not, in the case of a union, it's actually just raw memory in disguise. Passing the value by ref paints the exact wrong picture about what's going on there and further lends to the deception. `new()` simply does not accept values as arguments; it accepts memory as argument, and returns a brand "new" start-of-life value.

So, assuming `new()` receives a `void` of some kind (after sleeping on it some, I'm prepared to die on that hill), then your union initialisation looks like one of these cases:

If it receives `void*`, then obviously your union code will look like:
```d
new(&myUnion.member) T();
```
Where the pointer will naturally cast to `void*`. This expression is not not too annoying from a syntax perspective, but like you say, the type erasure loses the size information, so you can't do a bounds check. 

The solution that gives you bounds checking is to define `new(void[]) T`, and then your union initialisation will look like this:
```d
new((&myUnion.member)[0..1]) T();
```
So, I'm taking the address of the union member, and from that pointer taking a 1-length slice. That slice will naturally casts to `void[]` with the proper length... now the placement new will have the buffer size, and can perform the bounds checking you seek.

I grant that looks a tad awkward, but I think it's pretty much only the union case which is subject to this weird contortion, and my feeling is that it's appropriate to place that burden on the union case since they're already an unsavoury and weird case themselves. I would load the union case with the unnatural thunk, rather than requiring all the natural cases to be burdened with an unnatural thunk.

The natural cases look like this:
```d
T* newObject = new(allocateMem(T.sizeof)) T;
```

Given some allocator `void[] allocateMem(size_t bytes)`, which is how allocators in D are always naturally defined.

Even CRT malloc always seems to be wrapped in some function like `void[] d_malloc(size_t bytes) => malloc(bytes)[0..bytes]` everywhere I look; that's what's overwhelmingly established in D, so I reckon it's probably most appropriate to make the argument of new match the natural result of an allocator, and also by accepting `void[]`, you get bounds checking.

I think that's the way forward.

TurkeyMan commented:
Actually, thinking a little more, I realised one more really common case:
```d
  void opAssign(ref typeof(this) rh)
  {
    this.destroy();
    new(&this) typeof(this)(rh);
  }
```

It's common to fabricate a copy assignment from a copy constructor using that pattern. And so this is another case where `new(void[])` would feel awkward, but here's a wild idea...

It's always sat a little awkwardly with me that you have some T, you may destroy it and then pass that same T reference forward into a reconstruction... there's a period between destruction and reconstruction where you're handing a reference to raw and invalid memory, but it's still typed like a valid T.

From a purist perspective, it would be _nice_ if the language were specified that any time you are holding a T*, then you are holding a _valid and live_ value. That may even be a convenient piece of language spec in the event you were planning to implement lifetime analysis for instance.

Okay, so here's my idea; we have `void destroy(T*)`, which destroys some T... let's just change that to `void[] destroy(T*)`, which destroys T _and returns the buffer that used to contain the value back to the application_.

This actually feels quite natural to me; if you destroy a value, that terminates the value's lifetime, there is no more value and so there is no more `T*`, but there IS still the memory where T used to be... it makes sense for the result of destroy() to be the buffer where the value used to exist, and it's useful for the application to receive it, because it can be used for recycling.
Given that new definition for `destroy`, it makes the case I describe above convenient:

```d
T* object = ...;

void[] mem = destroy(object); // returns the former values buffer
T* newObject = new(mem) T; // recycle the buffer directly
```

Or simply:

```d
void opAssign(ref typeof(this) rh)
{
  new(this.destroy) typeof(this)(rh);
}
```

That reads nicely, the types all flow... but what's really nice is that because destroy returns the sized void buffer, you are no longer encouraged to handle the typed pointer post-destruction!
In the future with lifetime analysis; we can make it a compile error to handle the post-destroyed value, and insist that the user handle the untyped raw buffer returned from destroy instead.

Terrible idea?
I really like it.

After a call to destroy where you receive the value's former buffer as result; it is UB or even a compile error to touch the dead-but-still-typed pointer from that event onwards.


In the future; ideally, destroy would be intrinsic too, and not just call the destructor but also implement the termination of the value's lifetime. Maybe you should recycle `void[] delete(T*)` for an intrinsic destroy expression to complement placement new?

TurkeyMan commented:
> P.S. lifetime analysis of a pointer to an object is more difficult than lifetime analysis of an object, because of the possibility of multiple pointers to the same object.

Yeah, but tough; placement new is _FOR_ dealing with freely allocated objects. Your union case is an edge case, and it's also a language aberration.

For what it's worth; I find that I rarely use unions in my own code. I usually place an embedded buffer and implement emplace operations and casts in the methods that access what might have otherwise been different members in a union.
I only ever find myself use a union when I know I'm only dealing with POD, and in that case, why would I be concerned with placement new?

WalterBright commented:
> I only ever find myself use a union when I know I'm only dealing with POD, and in that case, why would I be concerned with placement new?

This code in dmd's constfold.d:
```
UnionExp Com(Type type, Expression e1)
{
    UnionExp ue = void;
    Loc loc = e1.loc;
    emplaceExp!(IntegerExp)(&ue, loc, ~e1.toInteger(), type);  // (*)
    return ue;
}
```
(*) would be:
```
    new(ue) IntegerExp(loc, ~e1.toInteger(), type);
```
It looks nice, and no ugly casting or use of &. Compile time checking of sizes, instead of runtime. The "joints" between untyped data and typed data are where user error creeps in. Minimizing this is good. This is why I want to try this approach.

I've made some progress in the dmd source with eliminating the use of the & operator. The results are appealing, more readable, and safer.

I know I haven't addressed all your points, more later.

TurkeyMan commented:
I forgot to respond to one of your points:

> 3. The lifetime of any pre-existing lvalue is going to be unceremoniously terminated when used by placement new, without passing Go or collecting $200 or running the destructor. This is why it is fundamentally unsafe. But at least refs can be statically checked for correct size!

I just want to point out here, that the idea of the lifetime of something being unceremoniously terminated is definitely not something that should be happening in any case ever. Cases of buffer recycling should have been destroyed prior.
This point you make is supporting my case that new should not accept `ref T` or `T*` as argument... by accepting `void[]`, then neither `ref T`, `T*`, or `Class` arguments can naturally convert to a `void[]`, so an accidental error is almost impossible.

The proposal I made above where destroy returns you the buffer that you need for a memory recycling operation feels like the natural way to deal with this situation; it naturally avoids this "unceremoniously terminating" live values, since they are not valid function arguments to begin with.

In fact, I would say that the pattern where destroy returns a `void[]` and new accepts a `void[]`, and consequently there are no blind casts or thunks in sight actually IS safe... with my proposal above, I think placement new (and placement delete) can both be safe operations.

TurkeyMan commented:
> > I only ever find myself use a union when I know I'm only dealing with POD, and in that case, why would I be concerned with placement new?
> 
> This code in dmd's constfold.d:
> 
> ```
> UnionExp Com(Type type, Expression e1)
> {
>     UnionExp ue = void;
>     Loc loc = e1.loc;
>     emplaceExp!(IntegerExp)(&ue, loc, ~e1.toInteger(), type);  // (*)
>     return ue;
> }
> ```
> 
> (*) would be:
> 
> ```
>     new(ue) IntegerExp(loc, ~e1.toInteger(), type);
> ```
> 
> It looks nice, and no ugly casting or use of &. Compile time checking of sizes, instead of runtime. The "joints" between untyped data and typed data are where user error creeps in. Minimizing this is good. This is why I want to try this approach.
> 
> I've made some progress in the dmd source with eliminating the use of the & operator. The results are appealing, more readable, and safer.
> 
> I know I haven't addressed all your points, more later.

I'm sorry, but using a union as primary (singular!) case-study is completely out of order.

You can write this:
```d
new((&ue)[0..1]) IntegerExp(loc, ~e1.toInteger(), type);
```

Yeah, it's not perfect, but neither are unions... don't wreck everything for this one degerate case.

My wall of text above should be seriously considered... I think it's a _really_ strong design, and solves a whole lot of problems that C++ (and D) has with lifetime flow, it makes the start/end of lifetimes feel natural while also making the typed variables naturally match the lifetimes of the values they represent, and it also solves convenient recycling of memory.
If you don't immediately understand the merit of my suggestion, then let's get on a phone call... I need an opportunity to represent the idea clearly, if it's not clear above.

WalterBright commented:
Given:
```
T* newObject = new(allocateMem(T.sizeof)) T;
```
how does this look:
```
T* newObject = new(mallocate!T()) T;
```
? Where:
```
ref T mallocate(T)()
{
    return *cast(T*)malloc(T.sizeof);
}
```
Not bad! Only one dirty cast for all the new's, no &, and the user needn't use a `sizeof`.

TurkeyMan commented:
1. I think that's the most insane permutation yet
2. I feel like you've dismissed everything I wrote :/

TurkeyMan commented:
This is what a function called `mallocate(T)()` is expected to do:
```d
void[] mallocate(size_t);

ref T mallocate(T)()
{
    return *new(mallocate(T.sizeof)) T;
}
```

No casts. It's completely @safe.

No function that returns a T should return an invalid/uninitialised T _ever_.

TurkeyMan commented:
My whole point is, there's an opportunity here to eliminate UB, invalid memory references, and unsafe casts associated with lifetime management from the language specification _entirely_... and that's a HUGE deal. If you haven't understood the point of what I'm suggesting, then we need to get on a call and talk.

WalterBright commented:
We both want the same thing. But realistically, there's no way to not have UB when allocating and initializing memory. That's why at some point it's going to have to not be verifiably safe by the compiler. As for void initialized objects,
```
S s = void;
```
we already have that in the language, and it is supported. But only in system code. Casting an array of void into a real object is always going to be system code. I appreciate your efforts to make this as safe as practical. But I also want the union case to be easy - this also applies to option types and sum types. And I want things to be correct without adding runtime checks. So here goes, I verified that it works:
```
import core.stdc.stdlib;

struct S {    int i = 1, j = 4, k = 9; }

ref void[T.sizeof] mallocate(T)() {
    return *(cast(void[T.sizeof]*) malloc(T.sizeof));
}

void main() {
    S* ps = new(mallocate!S()) S;
    assert(ps.i == 1);
    assert(ps.j == 4);
    assert(ps.k == 9);
}
```
I changed your proposal of a slice `void[]` to a static array `void[sizeof(T)]`, so the compiler knows its size at compile time. But does it blend? Let's check the code generated:
```
000c:   B8 01 00 00 00           mov       EAX,1
0011:   48 89 45 F8              mov       -8[RBP],RAX
0015:   E8 00 00 00 00           call      mallocate
001a:   48 89 C3                 mov       RBX,RAX              ; RBX gets pointer to storage
001d:   48 8B 45 F8              mov       RAX,-8[RBP]
0021:   89 03                    mov       [RBX],EAX            ; .i = 1
0023:   C7 43 04 04 00 00 00     mov       dword ptr 4[RBX],4   ; .j = 4;
002a:   C7 43 08 09 00 00 00     mov       dword ptr 8[RBX],9   ; .k = 9
```
Efficient, compact, and no monkey business.

TurkeyMan commented:
`return malloc(T.sizeof)[0 .. T.sizeof]` should be sufficient; you don't need that weird thunk.

This said though; you keep using `malloc()` as your experiment, but nobody in D calls malloc without wrapping it in a function that just returns a `void[]` instead of `void*`; all the memory allocators, (including Andrei's std API), all established long ago that we return `void[]` from allocation functions... this is why I'm trying to make it native compatible with that type.

And sorry, let me be more clear; yes I understand there will be unsafe escape hatches (union for instance, that's unsafe by definition), and some low-level coercions are still available and useful, but what I'm trying to achieve is where the common case (allocate and initialise) have no casts, no weird shit, so _that_ flow can actually be completely safe.


I thought about static array briefly too. It seems workable, but it's got some ergonomic problems. You can't pass fixed arrays easily, because you need a different overload of a function for every array size, which means this will always be wrapped in something that; receives a dynamic array, does a bounds check, and then slices the appropriately bounded static array, then pass to a call like you show above.

The nice thing about a fixed array is that you can elide runtime bounds checking, which is nice, but I think the ergonomic problem carries more weight. Memory allocation is NEVER a high-frequency operation; I can't imagine any situation where I wouldn't tolerate a runtime bounds check in the `void[]`.

Perhaps you'd consider accepting BOTH fixed array and also dynamic array; if a dynamic array was supplied, perform a runtime bounds check, static array, static bounds check?


I want to re-enforce though that my proposal is designed to work very nicely if we also make `destroy` return the buffer to the user (for recycling or returning to the allocator). The whole point is that in idiomatic and safe code, typed variables will never be handled at any time when they are in an invalid state. I really like that the type is only introduced to the program as a result of `new`; that is, the moment of its birth, and then you supply it to `delete` or `destroy` and it's gone; you get the void buffer back as a result which is the only thing you need to handle to perform a recycle, or just hand that memory back to the allocator. Your final contact with the typed value is supplying it as argument to `delete`.

WalterBright commented:
> return malloc(T.sizeof)[0 .. T.sizeof] should be sufficient; you don't need that weird thunk.

It doesn't work because it returns a `void[]`, and its size cannot be checked at compile time. The static array works, I tried it and looked at the assembler.

> Memory allocation is NEVER a high-frequency operation

Placement new has a major use case for unions, option types, and sum types, and these can be very high frequency. See CTFE, which I sped up quite a bit by using emplace on a stack allocated union. I've also written a Javascript interpreter - the operand stack is a critical high-frequency piece of code, and the operands get overwritten like any stack. As I've shown, the static array method generates optimal code, so the user doesn't need to hack around it.

> we return void[] from allocation functions

Not a problem, it's still castable to a pointer to a static array, since the length is known in advance.

> You can't pass fixed arrays easily, because you need a different overload of a function for every array size

I haven't implemented placement new for variable sized arrays yet. The proposed solution uses a template to take care of placement new expressions for types of fixed size. I don't think it is a big deal, it's just a few instructions.

> Perhaps you'd consider accepting BOTH fixed array and also dynamic array; if a dynamic array was supplied, perform a runtime bounds check, static array, static bounds check?

I expect a different implementation will be required for new'ing variable length arrays, meaning a `void[]`.

> The whole point is that in idiomatic and safe code, typed variables will never be handled at any time when they are in an invalid state.

Placement new is never going to be safe. Prefixing a destroy() means that void-initialized objects will also be destroyed, and that's UB for sure.

TurkeyMan commented:
> > return malloc(T.sizeof)[0 .. T.sizeof] should be sufficient; you don't need that weird thunk.
> 
> It doesn't work because it returns a `void[]`, and its size cannot be checked at compile time. The static array works, I tried it and looked at the assembler.

No, your code you wrote:
```d
ref void[T.sizeof] mallocate(T)() {
    return *(cast(void[T.sizeof]*) malloc(T.sizeof)); // <-- this is gross and blind cast/thunk is unsafe
}
```

You don't need to write that. Just write what I suggested instead; you don't need the thunk:
```d
ref void[T.sizeof] mallocate(T)() {
    return malloc(T.sizeof)[0 .. T.sizeof]; // <-- this is not gross and it is safe
}
```

> > we return void[] from allocation functions
> 
> Not a problem, it's still castable to a pointer to a static array, since the length is known in advance.

Don't cast, just slice it to the proper length.

> > The whole point is that in idiomatic and safe code, typed variables will never be handled at any time when they are in an invalid state.
> 
> Placement new is never going to be safe.

I don't see why you think this... it's definitely safe. The example code you wrote above is already safe, ie:
```d
struct S {    int i = 1, j = 4, k = 9; }

ref void[T.sizeof] mallocate(T)() {
    return malloc(T.sizeof)[0 .. T.sizeof]; // no casts, nothing nasty
}

void main() {
    S* ps = new(mallocate!S()) S;
    assert(ps.i == 1);
    assert(ps.j == 4);
    assert(ps.k == 9);
}
```

What's unsafe about this? It's completely fine by my reading.

> Prefixing a destroy() means that void-initialized objects will also be destroyed, and that's UB for sure.

I don't know what `prefixing a destroy` means, but you're scaring me... nothing weird please! I'm satisfied at this point that we're going the right direction here!

TurkeyMan commented on compiler/test/runnable/placenew.d:
```diff
@@ -0,0 +1,106 @@
+
+void test6()
+{
+    S6* ps = new(mallocate!S6()) S6;
```
This is the case we recently discussed... but what are all the other cases?
I thought we finally agreed on raw buffer initialisation using sized void buffers?

WalterBright replied:
Whether the void array buffers or used or a variable is up to the user. I left it in because it makes it most convenient when dealing with unions / optional types / sum types. The implementation doesn't care what type it is, only that it is an lvalue of sufficient size.

I didn't add tests for other cases, as they would be redundant. A tutorial should include them, however.

TurkeyMan commented on compiler/test/fail_compilation/placenew.d:
```diff
@@ -0,0 +1,80 @@
+void test0()
+{
+    int i;
+    int* pi = new (i) int;
```
I thought we ejected this pattern into space? Please, just no.

WalterBright replied:
I responded to this in the other comment.

Connor-GH commented:
This is code straight ripped from my kernel:

```
module memory;
import traits;
import kalloc : kmalloc, kfree;

T *d_new(T, Args...)(auto ref Args args) {
	import core.lifetime : emplace, forward;
	T *mem = cast(T *)kmalloc(T.sizeof);
	return mem.emplace(forward!args);
}
void d_delete(T)(T *data) {
	if (data is null)
		return;
	kfree(data);
}
```

I believe placement new would replace at least the d_new, but what about the d_delete? ``delete`` is deprecated, and in a betterC environment, I don't have access to classes and interfaces to even make an interface to the GC. In other words, I have *no* way of being able to use keyword ``new`` and keyword ``delete``.

WalterBright commented:
```
ref void[T.sizeof] mallocate(T)() {
    return malloc(T.sizeof)[0 .. T.sizeof]; // <-- this is not gross and it is safe
}
```
you're right, that is better. Thanks!

> What's unsafe about this?

It's equivalent to this:
```
struct T { int* a,b,c; }

void foo(T);

@safe void test(void[24] a) {
    foo(cast(T)a); // Error: cast from `void[24]` to `T` not allowed in safe code
}
```
I.e. the compiler cannot tell what it's stepping on, as it does not know what the provenance of the `void[]` is.

TurkeyMan commented:
Okay, so, what I'm suggesting by `new` accepting `void[]`(and only that), is that it DOES know; it can assume (written in the spec even), that the argument it receives is raw memory, and any argument that is not raw memory is an invalid argument. It can be confident that given raw memory, it has agency to construct a new value, and it is safe to do so. 

This assumption and assertion only works well when the function is specified to accept `void[]`.
This assumption gets awkward and tenuous if `new` accepts a `ref T`, because there's nothing stopping the user from passing a live or valid T, or distinguishing it from an uninitialised T; and `new` can't know. But if new only accepts `void[]`, then it knows it received raw memory, and the only way a user could pass something risky is with an explicit cast in user code, at which point the user has accepted an unsafe operation in support of their optimisation.

If the user has responsibility for the unsafe cast, then the operation itself is safe... this is quite specifically why I reject the idea that new would receive a T so vehemently; it fundamentally undermines the opportunity to make this whole thing safe, and wrecks lifetime analysis.

TurkeyMan commented:
I don't see how you get from `ref T` to `void[]` (or `void[N]`) though without unsafe casts, even considering the implicit cast from `T*` to `void*`. You need to write something like `(&lvalue)[0..1]`; taking the pointer is safe, but slicing a pointer to an array is not (pointer doesn't know the length); so the implicit cast to `void` array is not where it gets caught; in this expression, it's the blind slice on the `T*`.

> Not accepting T as an argument means the user will have to cast it to void*, which will look ugly. Casting to void* is a safe operation, so it becomes an ugly construct with no added value.

I don't think it looks ugly, it looks appropriate. It's visible to the reader that you're taking the buffer from under a value; which is what you are doing. I think the clarity of that operation holds huge value.

And again, you're arguing from niche-cases; the default case that people will type doesn't encounter any of this; the 2 cases you've demonstrated that do are your union (unsafe by definition; special handling is completely appropriate), and a `void`-init stack variable; which is fairly uncommon; in almost all circumstances you just let the declaration init the value... but a `void`-init stack variable is also unsafe, and requires handling a typed value before it's life. Maybe the programmers in the safe-future will see a shift to placing buffers on the stack rather than uninitialised lvalues.

I think it's fundamentally bad for the language at the lowest level to promote or require handling invalid-but-already-typed objects. If the variable hasn't been initialised yet, it's memory, so it's `void[]`.

Talking about option types and sum types is a red-herring; I use option types and sum types too, but they are a tool in the toolbox. You don't write placement new ever while using those objects; they have strong API associated with initialisation and assignment. Everything's internal, nobody will handle placement new expressions when interacting with those objects.

Optimise for what people actually do in their code, which is allocate memory and then initialise it.

> Note there's an inconsistency in the compiler - in @safe code, a T cannot be cast to a void[T.sizeof], but an &T can be cast to void*. https://issues.dlang.org/show_bug.cgi?id=24866

Is that an inconsistency? That's specifically what I'm taking advantage of.
It's never occurred to me to close that gap that way... I quite like that how it is.

> The ref T semantics was chosen because:

They weren't "chosen"; you just trivially decided it with no warning or discussion, against the strongest advice I can possibly muster.

> 1. under the hood a simple pointer is passed - this is the most efficient code
> 2. the size of the object pointed to by the pointer is known at compile time and can be checked at compile time

`void[N]` can be passed by ref, and the size is known at compile time.

> Such efficiency matters a great deal with unions, option types, and sum types. We have to be competitive.

Complete red herring; I'm not suggesting anything that would change the codegen in any way whatsoever; I'm interested in getting the expression right; clear, self-explanatory, and relatively foolproof. Nothing I'm discussing changes a single thing about the codegen.

I have proposed the OPTION to accept `void[]` in ADDITION to `void[N]`, which would causes a runtime bound check to be emit, but if that doesn't land, I'm happy with `void[N]` only; the user will have to slice to `N` length, and cause the runtime bounds check at the users slice expression rather than the `new` expression.

WalterBright commented:
> I have proposed the OPTION to accept void[] in ADDITION to void[N]

I avoided that because if `T` is a `void[]` then the compiler will have to introduce a special case to have T not be the target but what it refers to being the target.

> Talking about option types and sum types is a red-herring

I use emplace 206 times in the compiler front end. I use a more primitive method extensively in the optimizer and back end. It's also used extensively in Phobos.

We're just going in circles now repeating ourselves.

The good news is that it works just fine for your use case. You can get rid of emplace now! I'm a bit less lucky, as we're still using an ancient version of D for the bootstrap compiler, grr grr.

> you just trivially decided it with no warning or discussion

This is a bit unfair. If Bob writes a proposal, he makes numerous decisions. People then ask for a reference implementation. Bob implements it, discovering that more decisions have to be made. Then the discussion happens. That's normally how things work.

TurkeyMan commented:
> > > I have proposed the OPTION to accept void[] in ADDITION to void[N]
> > 
> > 
> > I avoided that because if `T` is a `void[]` then the compiler will have to introduce a special case to have T not be the target but what it refers to being the target.
> 
> I don't follow. I don't see how there's any room for special cases here?

I realised what you meant here... this situation naturally occurs when you want to new an array though, so the case still has to be handled?
Why do you see this case as a problem?

WalterBright commented:
I am catching up after 4 days without internet or power. Seattle is a technologically advanced city with a primitive electric grid.

 > I consistently feel casually dismissed

I'm sorry you got that impression. I've provided a rationale for each decision.

> Should I have perceived that it was forked from my initial proposal and I was cut out of it? I'll admit that I was somewhat offended to note my name's nowhere to be seen near the DIP on this. That feels a bit unfair... but perhaps that's the signal where I should have recognised that it's no longer my design.

You did indeed propose `new(ptr) T(...)` in email along with a couple examples.

The first line of the first post of the DIP in the n.g. on Oct 30: "Based on a suggestion by Manu Evans:"
https://www.digitalmars.com/d/archives/digitalmars/dip/development/First_Draft_Placement_New_Expression_421.html#N421
I updated the DIP Nov 18: https://github.com/WalterBright/documents/blob/master/placementnew.md

The dip is about passing an lvalue rather than a pointer.

> perverting ... butchering

It's a good thing we're friends, Manu!

If you want things verbatim, write the DIP the way you want it. I wrote it the way I thought would work best. I've never seen an idea that survived into a specification without modification, like the replacement of void* with void[]. I've also never seen a specification that survived into an implementation without modification. (I rewrote the DIP after this implementation.)

You've wanted 3 things, and I want them too:

1. __rvalue
2. placement new
3. extended alias

The first two now have implementations and will work for your use cases. The other aspect of placement new works for my use cases. We should be both happy about this.

> but there was a lot more to it that I couldn't easily share in the initial conversation we had.

I didn't see the more, other than what I mentioned earlier. I know placement new from C++, having implemented it. I also implemented that in an earlier version of D: https://dlang.org/deprecate.html#Class%20allocators%20and%20deallocators and the implementation code for it is still present (and I made some use of it!).

> From my point of view; this is my proposal...

Ok, but you've also made it clear it is not what you proposed :-)

The hinge of our disagreement is placement new cannot be made `@safe` and trying to make it safe with an unsafe cast doesn't fix it. The implementation won't allow its use in `@safe` code.

TurkeyMan commented:
> The hinge of our disagreement is placement new cannot be made `@safe` and trying to make it safe with an unsafe cast doesn't fix it. The implementation won't allow its use in `@safe` code.

Not at all, the hinge of our disagreement is that accepting a `T` as input is absolutely unacceptable.
Accepting a `T` _by `ref`_ is palpably worse, because it further hides the appearance that you're handling memory at all (no pointer in sight) and looks exactly like passing a value.

By writing:
```d
T x;
new(x) T;
```
There's nothing you could do to make it look more like a value of `T` is the input to the function, which couldn't be further from the truth; the ONLY valid input is raw uninitialised memory. The only language we have to handle that concept is `void`.

I think if you were going to design this API, it's not possible to make the API worse than that; it communicates exactly the wrong thing, and I don't think you could inspire misunderstanding and user error more confidently if you tried.

Before `new`, there IS NO `T`... I'll die on that hill.
The language should not have any spec that encourages handling invalid values, and I reckon it will also complicate future lifetime analysis.

Yes I know you can write `T x = void` and then you want to hand that to placement new; I absolutely want to see the coersion from `T` to `void[N]` right in my face, so that there can be no mistake. Anyone that encounters that line of code will stop and ask themselves if the value they're coercing has definitely been destroyed prior, or assure it's uninitialised.

nordlow commented:
Could the current semantics of C++'s " Placement new" described in https://en.cppreference.com/w/cpp/language/new give insights on what type(s) that should be allowed to be passed as argument to `new()`?

TurkeyMan commented:
The trouble with the function apparently accepting `T` values, is that people will supply `T` values! They will get their new object out the other side like they expect, and (depending on the implementation of the constructor/destructor) there's a high chance their code will appear to just work. There's nothing in sight to suggest they might be making a mistake, and chances are they are just silently leaking memory or something like that.

TurkeyMan commented:
Oh no!
```
Error: cannot use `new` in `@nogc` function `...`
```

Seems that placement new is not `@nogc`, probably hangover from `new`?
The expression needs to infer the attributes of the constructor it calls...

TurkeyMan commented:
```d
    override void visit(NewExp e)
    {
+++     if (e.placement)
+++         return; // placement new
        if (e.member && !e.member.isNogc() && f.setGC(e.loc, null))
        {
            // @nogc-ness is already checked in NewExp::semantic
            return;
        }
        if (e.onstack)
            return;
        if (global.params.ehnogc && e.thrownew)
            return;                     // separate allocator is called for this, not the GC

        if (setGC(e, "cannot use `new` in `@nogc` %s `%s`"))
            return;
        f.printGCUsage(e.loc, "`new` causes a GC allocation");
    }
```

This worked for me locally...

WalterBright commented:
@TurkeyMan  Added your fix

TurkeyMan commented:
Is it correct though?

WalterBright commented:
yes

TurkeyMan commented:
shall we merge this too?

TurkeyMan commented:
Actually, I still vehemently object to the API... can we get some 3rd parties to add their opinions?

atilaneves commented:
Let me talk to @WalterBright first.

WalterBright commented:
Documentation PR: https://github.com/dlang/dlang.org/pull/4161

WalterBright commented:
Note: I have not added the `T[]` placement expression yet. I figure this PR is large enough without adding that in. Will do in a followup.

thewilsonator commented:
I presume this is no longer WIP? If so please remove the label

TurkeyMan commented:
Well, I clearly lost the debate on the lvalue matter... so other than that, is there any reason this is left hanging?
I'd like to start working it into the library...

This patch LGTM. Anyone care to approve or merge?

thewilsonator commented:
Needs a rebase though.

TurkeyMan commented:
I don't understand these CI failures; what's going on here? I see a lot of what look like spurious failures; is the CI generally unstable?

thewilsonator commented:
FreeBSD is known to time out, Alpine I'm not sure.

TurkeyMan commented:
How do you issue a re-run?

TurkeyMan commented:
These are all weird CI errors, is it unusual to see this sort of thing, or is there some chance that this is related to the PR itself in some way? It's pretty weird. I wish I could click re-run on those specific tests without restarting the whole thing... which I can only see how to do with another force-push...?

TurkeyMan commented:
I think it was just an anomaly; I re-ran the failed tests and they completed in a 'normal' time-frame.
I'm-a click the button...

dkorpel commented:
The DIP has not been reviewed or formally assessed yet, so I think it's a little early to merge this without at least preview switch. Especially since this clashes with primary type syntax, a DIP which is being formally assessed right now.

TurkeyMan commented:
Is there a DIP? Walter merged the move constructor stuff already, and this is related to that. I can't make progress on the move semantics work without this...

Herringway commented:
The DIP process should really just be cancelled if nobody's going to respect it. What a waste of everyone's time.
